misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt

```diff
@@ -0,0 +1,109 @@
+package misk.metrics.digester
+
+import misk.metrics.Histogram
+import misk.metrics.HistogramRecordMetric
+import misk.metrics.HistogramRegistry
+import java.util.Objects
+import java.util.concurrent.ConcurrentHashMap
+
+/** DigestMetric contains the contents of an individual metrics within a histogram */
+class DigestMetric(
+  val digest: SlidingWindowDigest<*>,
+  val labels: List<String>
+) {
+  /** Adds an observation to the metric */
+  fun observe(value: Double) {
+    digest.observe(value)
+  }
+}
+
+class TDigestHistogramRecordMetric constructor(private val metric: DigestMetric) :
+    HistogramRecordMetric {
+  override fun observe(duration: Double) {
+    metric.observe(duration)
+  }
+}
+
+/** TDigestHistogram stores histograms which records metrics */
+class TDigestHistogram<T : TDigest<T>> constructor(
+  val name: String,
+  val help: String,
+  val quantiles: List<Double>,
+  private val labelNames: List<String>,
+  private val tDigest: () -> SlidingWindowDigest<T>
+) : Histogram {
+
+  private val metrics = ConcurrentHashMap<Int, DigestMetric>()
+
+  /** Add all labels passed to the histogram */
+  init {
+    labelNames.forEach { label ->
+      labels(label)
+    }
+  }
+
+  /** Records a new metric within the histogram */
+  override fun labels(vararg labelValues: String): TDigestHistogramRecordMetric {
+    val metric = metrics.getOrPut(key(labelValues)) {
+      DigestMetric(tDigest(), labelValues.asList())
+    }
+    return TDigestHistogramRecordMetric(metric)
+  }
+
+  /** Returns the number of windows within the histogram */
+  override fun count(vararg labelValues: String): Int {
+    return metrics.get(key(labelValues))?.digest?.windows?.count() ?: 0
+  }
+
+  /** Returns a metric of the histogram. Order of labels matters */
+  fun getMetric(vararg metric: String): DigestMetric {
+    return metrics.getValue(key(metric))
+  }
+
+  /** Creates a hash for a list of labels */
+  private fun key(labels: Array<out String>) = Objects.hash(labels.sorted())
```

Comment:
Don't sort here. This is sorting label _values_, which you definitely don't want to do.

The Go implementation is based on a map, which is why it sorts based on the map keys.

Reply:
Don't we want labels ["label1",label2"] to have the same key as ["label2",label1"]?

Reply:
No, since the key is based on label _values_ and not label names. So for this you need `Objects.hash(*labels)` (note the important `*` to expand the varargs).

Each `TDigestHistogram` instance represents a metric with a fixed name, a set of labels with fixed names and differing values. They key in the metrics map for this class represents a unique combination of ordered label values.

Let's say there's an instance named `hist` of `TDigestHistogram` with `labelNames == ["a-label", "b-label"]`. Calling
`hist.labels("x", "y")` and `hist.labels("y", "x")` creates two different metric values in Prometheus:
1. `{"a-label":"x", "b-label":"y"}` where the key in your metrics map is `Objects.hash("x", "y")`
2. `{"a-label":"y", "b-label":"x"}` where the key in your metric map here is `Objects.hash("y", "x")`

Even though both examples use `x` and `y` as values for labels, they use them as values for different label names. `Objects.hash("x", "y") != Objects.hash("y, "x")`. If you were to sort label values, they two metric values would end up being incorrectly squished into one metric value.

Reply:
This still can't be sorted. Let's talk offline to clarify.